opennebula/resource_opennebula_virtual_network.go

```diff
@@ -501,215 +493,121 @@ func generateAR(armap map[string]interface{}, id int) string {
-			return fmt.Sprintf(arstring, arid, armac, arsize)
-		}
-		var arstring = `AR = [
-            AR_ID = %s,
-		    TYPE = ETHER,
-		    SIZE = %d ]`
-		return fmt.Sprintf(arstring, arid, arsize)
-	}
-
-	return ""
+	ar.Add(vnk.ARID, arid)
+	ar.Add(vnk.Size, fmt.Sprint(arsize))
+	ar.Add(vnk.Type, artype)
+
+	if armac != "" {
+		ar.Add(vnk.Mac, armac)
+	}
+
+	switch artype {
+	case "IP4":
+		ar.Add(vnk.IP, arip4)
+
+	case "IP6":
+
+		if argprefix != "" {
+			ar.Add(vnk.GlobalPrefix, argprefix)
+		}
+
+		if arulaprefix != "" {
+			ar.Add(vnk.UlaPrefix, arulaprefix)
+		}
+
+	case "IP6_STATIC":
+
+		ar.Add("IP6", arip6)
+		ar.Add(vnk.PrefixLength, arprefixlength)
+
+	case "IP4_6":
+
+		if argprefix != "" {
+			ar.Add(vnk.GlobalPrefix, argprefix)
+		}
+
+		if arulaprefix != "" {
+			ar.Add(vnk.UlaPrefix, arulaprefix)
+		}
+
+		ar.Add(vnk.IP, arip4)
+
+	case "IP4_6_STATIC":
+
+		ar.Add(vnk.IP, arip4)
+		ar.Add("IP6", arip6)
+		ar.Add(vnk.PrefixLength, arprefixlength)
+	}
+
+	return ar.String()
 }
 
 func generateVnTemplate(d *schema.ResourceData) (string, error) {
+
+	tpl := vn.NewTemplate()
+
 	mtu := d.Get("mtu").(int)
-	dns := d.Get("dns").(string)
-	gateway := d.Get("gateway").(string)
-	netmask := d.Get("network_mask").(string)
-	description := d.Get("description").(string)
 	guestmtu := d.Get("guest_mtu").(int)
 
 	if guestmtu > mtu {
 		return "", fmt.Errorf("Invalid: Guest MTU (%v) is greater than MTU (%v)", guestmtu, mtu)
 	}
 
-	vntpl := &vnetTemplate{
-		Description:  description,
-		Mtu:          mtu,
-		GuestMtu:     guestmtu,
-		Dns:          dns,
-		Gateway:      gateway,
-		Network_Mask: netmask,
-	}
-
-	w := &bytes.Buffer{}
+	tpl.AddPair("MTU", mtu)
+	tpl.AddPair(string(vnk.GuestMTU), guestmtu)
 
-	//Encode the VN template schema to XML
-	enc := xml.NewEncoder(w)
-	//enc.Indent("", "  ")
-	if err := enc.Encode(vntpl); err != nil {
-		return "", err
+	if dns, ok := d.GetOk("dns"); ok {
+		tpl.Add(vnk.DNS, dns.(string))
+	}
+	if gw, ok := d.GetOk("gateway"); ok {
+		tpl.Add(vnk.Gateway, gw.(string))
+	}
+	if netMask, ok := d.GetOk("network_mask"); ok {
+		tpl.Add(vnk.NetworkMask, netMask.(string))
+	}
+	if desc, ok := d.GetOk("description"); ok {
+		tpl.Add("DESCRIPTION", desc.(string))
 	}
 
-	log.Printf("template XML: %s", w.String())
-	return w.String(), nil
+	tplStr := tpl.String()
+	log.Printf("[INFO] VNET template: %s", tplStr)
+
+	return tplStr, nil
 }
 
-func generateVnXML(d *schema.ResourceData) (string, error) {
+func generateVn(d *schema.ResourceData) (string, error) {
 	vnname := d.Get("name").(string)
 	vnmad := d.Get("type").(string)
 
 	if vnmad == "" {
 		vnmad = "bridge"
 	}
-	vnstring := fmt.Sprintf("NAME = %s\nVN_MAD = %s", vnname, vnmad)
+
+	tpl := vn.NewTemplate()
+
+	tpl.Add(vnk.Name, vnname)
+	tpl.Add(vnk.VNMad, vnmad)
 
 	if validVlanType(vnmad) >= 0 {
 		if d.Get("automatic_vlan_id") == true {
-			vnstring = fmt.Sprintf("%s\nAUTOMATIC_VLAN_ID = YES", vnstring)
+			tpl.Add(vnk.AutomaticVlanID, "YES")
```

Comment:
this key does not exist in the last 5.10 goca
Can you add explicitely the correct key `AUTOMATIC_VLAN_ID` instead ? Until the fix is available in OpenNebula 5.10 ?